### PR TITLE
AFK Rewards

### PR DESCRIPTION
**Pull request by lotuuu**
Adds AFK rewards.

In Champions of Mirra we have this mechanic called AFK rewards, where beating levels will grant you a different afk rate for currencies. These currencies will be accumulated and whenever you send a `claim` request, you will get whatever you had accumulated since last time you claimed (max. 12 hours).

# Known issues 

We should create new afk reward rates if we give increment_afk_reward_rate a currency that's not owned. Issue: https://github.com/lambdaclass/champions_of_mirra/issues/223

Exploitable bug when accumulating time and advancing levels before claim. Issue: https://github.com/lambdaclass/champions_of_mirra/issues/198

# To test

Create user, get campaigns, get levels, get currencies
```
{:ok, user} = Champions.Users.register("User")
{:ok, [campaign | _]} = Champions.Campaigns.get_campaigns
[level1, level2, level3 | _] = campaign.levels |> Enum.sort_by(& &1.level_number)
gold = GameBackend.Users.Currencies.get_currency_by_name!("Gold")
gems = GameBackend.Users.Currencies.get_currency_by_name!("Gems")
```

Check you have rates and they are `0.0`
```
GameBackend.Rewards.get_afk_reward_rate(user.id, gold.id)
GameBackend.Rewards.get_afk_reward_rate(user.id, gems.id)
```

Beat level 1 (gives only currency)
```
Champions.Battle.fight_level(user.id, level1.id)
```

Check out your gold (should be 110 from the sample currencies in user register + first level beaten)
```
GameBackend.Users.Currencies.get_amount_of_currency(user.id, gold.id)
```

Check your rates have now grown.
```
GameBackend.Rewards.get_afk_reward_rate(user.id, gold.id)
GameBackend.Rewards.get_afk_reward_rate(user.id, gems.id)
```

Wait a few moments and check how much you have accumulated to claim
```
Champions.Users.get_afk_rewards(user.id)
```

Claim it!
```
Champions.Users.claim_afk_rewards(user.id)
```

Check they have been consumed
```
Champions.Users.get_afk_rewards(user.id)
```

And that you have them in your UserCurrencies now!
```
GameBackend.Users.Currencies.get_amount_of_currency(user.id, gold.id)
```

### CODE REVIEW

File: apps/game_backend/priv/repo/migrations/20240219192225_add_experience_and_afk_rewards.exs
Comment (ncontinanza) [inline apps/game_backend/priv/repo/migrations/20240219192225_add_experience_and_afk_rewards.exs:1]: Maybe we could rename this migration to 2024..._add_afk_reward_rates to reflect better the changes that it implies.

**Review by ncontinanza — CHANGES_REQUESTED** [review]

**Review by ncontinanza — APPROVED** [review]

File: apps/game_backend/lib/game_backend/campaigns.ex [hunk 1]
```diff
@@ -19,7 +19,7 @@ defmodule GameBackend.Campaigns do
         )
       )
 
-    if Enum.empty?(campaigns), do: {:error, :no_campaigns}, else: {:ok, campaigns}
+    if Enum.empty?(campaigns), do: {:error, :no_campaigns}, else: {:ok, Enum.sort_by(campaigns, & &1.campaign_number)}
```
Comment (agustinesco) [inline apps/game_backend/lib/game_backend/campaigns.ex:22]: Isn't this `order_by: [asc: c.campaign_number]` already sorting the campaigns?

File: apps/game_backend/lib/game_backend/users.ex [hunk 2]
```diff
@@ -119,12 +121,25 @@ defmodule GameBackend.Users do
...
 
+  def reset_afk_rewards_claim(user_id) do
+    {:ok, user} = get_user(user_id)
+
+    user
+    |> User.changeset(%{last_afk_reward_claim: DateTime.utc_now()})
+    |> Repo.update()
+  end
+
   defp preload(user),
```
Comment (agustinesco) [inline apps/game_backend/lib/game_backend/users.ex:132]: we should change the name of this function in the future

**Review by agustinesco — APPROVED** [review]
LGTM!

**Pull request merged**